### PR TITLE
Migrate DoResult/EstimandResult internal storage to xarray

### LINKED ISSUE #319: Add EstimandResult for ate/cate/att/atu and __repr__ for DoResult
## Problem

`ate()`, `cate()`, `att()`, `atu()` all return `DoResult`, which has no `__repr__` and requires the user to re-specify the outcome variable to get a number (`result.mean("Y")`). This is confusing for new users (#315) and inconsistent with every other Result type in pathmc, all of which have `__repr__`.

### Two result types, one per intent

**`DoResult`** (returned by `do()`) — "what happens to the whole system under this intervention?" No single privileged outcome. Holds draws for all variables.

**`EstimandResult`** (new, returned by `ate()`, `cate()`, `att()`, `atu()`) — "what is the effect of X on Y?" Knows its outcome variable. Holds the same draws but provides a focused API.

Both hold `dict[str, np.ndarray]` internally. The difference is that `EstimandResult` knows what question was asked.

### What changes

**`DoResult` gets a `__repr__`:**
```
DoResult (2000 draws, 5 variables)
              mean   94% HDI
  X           1.00   [1.00, 1.00]
  M           3.45   [2.81, 4.12]
  Y           7.21   [5.93, 8.44]
```

**`EstimandResult` is a new class with:**

```python
result = m.ate("Y", "X", values=(0, 1))

# Notebook repr:
#   ATE of X on Y
#     Mean:    1.23
#     94% HDI: [0.87, 1.61]
#     P(> 0):  0.97
#     Draws:   2000

result.mean()              # defaults to outcome — no arg needed
result.hdi()               # defaults to outcome
result.draws()             # np.ndarray (n_samples,) for outcome
result.prob("> 0")         # P(ATE > 0)
result.summary()           # tidy DataFrame row
float(result)              # posterior mean via __float__

result.mean("M")           # still works for other variables
```

### Per-unit draws (stretch goal)

Currently, draws are averaged over units during `do()` computation. Instead, store the full tensor internally (xarray with named dims: `sample`, `unit`, optionally `time`) and summarize lazily. Default accessors give population-level summaries. An opt-in accessor exposes unit-level detail:

```python
result.draws()                  # (n_samples,) — population
result.draws(per_unit=True)     # (n_samples, n_units) — unit-level
```

### Implementation notes

- `prob()` uses an explicit method, not operator overloading — `result > 0` returning a float breaks Python `__gt__` semantics
- `__sub__` on two `EstimandResult`s with the same `_default_var` should preserve it
- All internal aggregations (e.g. averaging over units) must use named xarray dims (`.mean(dim="unit")`) not numerical axes (`.mean(axis=2)`) — the current numpy axis-based approach is fragile and hard to read

## Tasks

- [x] Add `__repr__` to `DoResult`
- [x] Create `EstimandResult` class with `__repr__`, `__float__`, `mean()`, `hdi()`, `draws()`, `prob()`, `summary()`
- [x] Wire `ate()`, `cate()`, `att()`, `atu()` to return `EstimandResult`
- [ ] Refactor internal draw storage to use xarray with named dims (`sample`, `unit`, `time`) instead of raw numpy with numerical axes
- [ ] Store full per-unit draws internally, add `per_unit`/`by_unit()` accessor
- [x] Update all docs pages that reference `DoResult` or show `ate()`/`cate()` usage (user guide + examples)
- [x] Highlight the new `prob()` method in the causal inference user guide
- [ ] Local re-freeze of docs by @drbenvincent after all changes land

## Related

- #315 — user confusion that prompted this

### PR DESCRIPTION
**Pull request by drbenvincent**
## Summary

Migrates the internal draw storage of `DoResult` and `EstimandResult` from `dict[str, np.ndarray]` to an `xarray.Dataset` with named dims (`chain`, `draw`, and optionally `unit`, `time`). Exposes the store publicly via a `dataset` property on both result classes. This is the last unchecked **code** task from #319.

Closes #319 (code). #315 closes as its dependency.

Towards #351 (API tests) — this PR adds substantial API test coverage as a side effect of the refactor safety net.

Follow-up tracked in #353: producers build `Dataset` directly and narrow dict constructors (tier 3).

### `pathmc/simulate.py` — the core refactor
- `DoResult` and `EstimandResult` hold `self._ds: xr.Dataset` with named dims `chain`, `draw`, and optionally `unit` (predictive path) and `time` (panel).
- **`dataset` property** — public, read-only access to the labelled `xr.Dataset` (ArviZ-style naming). `draws()` / `mean()` / `hdi()` remain the stable flat-numpy contract; `dataset` is the power-user escape hatch for `.sel()`, per-chain analytics, etc.
- **`_draw(var)` / `_draw_by_time(var)`** — O(1) single-variable stacking; all public accessors and internal callers route through these instead of rebuilding dict views.
- **`_build_dataset`** — single construction point from flat numpy dicts (test fixtures and transitional callers). Requires explicit `n_chains`/`n_draws`; raises on unrecognised lengths (no `sample_{var}` fallback). Predictive arrays use a shared `unit` dim via `n_units_per_var`.
- **`__sub__`** — element-wise contrast via xarray on shared variables with matching schemas (`lhs[common] - rhs[common]`).
- Producers (`run_do_pymc`, `run_do_panel_unified`) pass `n_chains`/`n_draws`/`n_units_per_var` so dims are labelled correctly.
- Panel per-time vars are stored once with a `time` dim; the cross-sectional `draws(var)` view averages over `time`.

### `pathmc/_model.py`
- `sensitivity()` and `prob()` updated to use `_draw()` / `dataset` instead of removed `_values` dict views.

### Tests
- **`tests/test_estimand_result_api.py`** — characterization tests for `EstimandResult` methods: `summary()`, `prob()`, `__float__`, `__sub__`, `from_contrast`, default-to-outcome resolution, HDI widening.
- **`tests/test_panel_by_time.py`** — characterization for `by_time()`/`time_index` on both classes, contrast `by_time` propagation, and `TestTimeDrawCount` guardrail.
- **`tests/test_xarray_storage.py`** — migration tests asserting `dataset` is `xr.Dataset` with named dims, coord alignment, numpy public outputs, xarray `__sub__`, `from_contrast` sharing, and chain/draw requirements.

## Verification
- Full test suite passes (`make test-fast` / `make test`).
- `make lint` clean (ruff, ruff-format, mypy, prek).
- The `TestDrawCount` guardrail still passes.

## Out of scope (per plan)
- **Per-unit draws / `by_unit()` accessor** — future PR; the `unit` dim is in place to make this straightforward.
- **`kind="predictive"` `flatten()` semantics** — latent inconsistency where predictive draws flatten units into the draw axis; flagged as a separate issue, not fixed here.
- **Tier 3 (#353)** — build `Dataset` in producers, narrow dict constructors, shared mixin, shrink/remove `_build_dataset`.
- **Docs re-freeze** — post-merge task.

## Design notes
- `dataset` returns the backing store by reference; `EstimandResult.from_contrast` shares it with the source `DoResult` contrast.
- Dict constructors remain for tests and transitional callers but require explicit chain/draw metadata; production paths already supply this via producers.

### CODE REVIEW

**Comment by drbenvincent** [comment]
@bugbot review

**Review by daimon-pymclabs — COMMENTED** [review]
## Review — xarray draw-storage migration

Solid refactor with a genuinely strong safety net (76 new tests, 0 existing tests modified — the best possible signal that the public contract held). The dict-view shim is a clean way to keep `_model.py` callers untouched. A few things worth addressing before merge, roughly in priority order.

### 1. Performance regression: `_values` rebuilds the whole dict on every access
`draws()`, `mean()`, `hdi()`, and `prob()` all still do `self._values[self._resolve(var)]` (simulate.py ~290–340). `_values` is now a **property** that stacks + transposes *every* variable into numpy on each call. So a single-variable `draws(var)` went from an O(1) dict lookup to O(n_vars × stack), and `summary()`/`prob()` call it more than once, compounding it. For predictive results (chain×draw×unit) that's a non-trivial allocation each time.

Suggest adding a private `_draw(var)` that stacks only `self._ds[var]` and routing the public accessors through it; keep `_values` as a compatibility shim purely for the `_model.py` external callers.

### 2. Heterogeneous unit counts can crash `_build_dataset` (or silently drop the `unit` dim)
The predictive path passes a single scalar `n_units = max(n_units_per_var.values())` (simulate.py ~965). In `_build_dataset`, a var whose unit count is neither `1` nor the max matches neither `nc*nd` nor `nc*nd*n_units`, so it falls to the legacy `dims=("sample",)` branch. Two such vars with *different* lengths then collide → `xr.Dataset` raises **"conflicting sizes for dimension 'sample'"**. Even when it doesn't crash, those vars silently lose the `unit` dim — which defeats the stated goal of a labelled `unit` dim for a future `by_unit()`.

Is a mixed-N across predictive vars reachable (e.g. `subgroup_indices` slicing applied to some vars but not others, or 3-D `var` vs 2-D `mu_var`)? If yes, this is a real gap. Recommend threading a **per-var** unit count (dict) into `_build_dataset` rather than a single max, and giving each fallback var a unique dim name.

### 3. `__sub__` switched from positional to coordinate alignment
Old code subtracted flat arrays element-wise (positional pairing); new code does `self._ds[common] - other._ds[common]`, which xarray aligns on the `chain`/`draw`/`unit` **index coords**. Because both sides use `np.arange` coords this is equivalent *today* — but it's a latent behavior change: if the two operands ever had mismatched lengths, the old code raised, the new code silently inner-joins (drops/NaN-fills). Worth a short comment documenting the assumption, plus a test asserting the mismatched-coord case behaves as intended.

### 4. Possible inconsistency between `_stack_sample_time` and the time branch of `_build_dataset`
`_stack_sample_time` explicitly handles a `unit` dim ("flattened into the sample axis (predictive panel)"), but the `values_by_time` branch of `_build_dataset` only ever reshapes to `(time, chain, draw)` and never creates a `unit` dim. So either (a) predictive-panel per-time arrays are unit-inflated, in which case `arr.reshape(n_times, n_chains, n_draws)` will raise on the size mismatch, or (b) they aren't, in which case the `unit` handling in `_stack_sample_time` is dead code. Either way it reads as an unresolved inconsistency — please confirm which path is real and prune/fix accordingly.

### 5. Test coverage gaps
- The `test_estimand_result_api.py` / `test_panel_by_time.py` fixtures build with `values=` and **no `n_chains`**, so they exercise only the `n_chains=1` fallback. The non-trivial multi-chain reshape→stack ordering is asserted for *shape* in `test_xarray_storage.py` but never for *value ordering*. Since `__sub__`'s positional pairing depends on that ordering being stable, add one explicit round-trip: known array → `DoResult(n_chains=2, n_draws=…)` → `draws()` equals the input in order.
- No test covers the mixed-length / heterogeneous-unit case from point 2.

### Minor
- Docstrings still say `values` is shape `(n_samples,)` with `n_samples = chain * draw`, but predictive vars are stored as `chain * draw * unit`. Tighten the wording so the contract isn't misleading.

Nice touches: the `TestTimeDrawCount` guardrail mirroring `TestDrawCount`, fully MCMC-decoupled hand-built fixtures, and keeping `_build_dataset` as the single construction point (dropping the redundant Phase 3 helper) are all the right calls.

**Comment by drbenvincent** [comment]
Addressed review feedback in 6e23ae3:

**1. `_values` performance** — Added `_draw(var)` / `_draw_by_time(var)` on both result classes. Public accessors (`draws`, `mean`, `hdi`, `prob`, `summary`, `by_time`) now stack a single variable instead of rebuilding the full `_values` dict. The dict-view properties remain for `_model.py` compatibility.

**2. Heterogeneous unit counts** — `_build_dataset` accepts `n_units_per_var`; `run_do_pymc` passes the full per-var dict. Vars with different predictive unit counts get per-var `unit_{var}` dims; legacy fallback arrays use `sample_{var}` to avoid Dataset collisions.

**3. `__sub__` alignment** — New `_subtract_stores()` subtracts stacked flat numpy arrays positionally and raises `ValueError` on length mismatch (no silent xarray inner-join). Tests cover legacy-vs-labelled layout and mismatched lengths.

**4. `_stack_sample_time` / panel units** — Removed dead `unit` handling; panel producers average over units before storage, so per-time arrays are always `(time, chain, draw)`. Documented in the docstring.

**5. Test coverage** — Added round-trip value ordering (`np.arange` through multi-chain reshape), heterogeneous unit counts, and subtraction edge cases.

**Minor** — Docstrings now distinguish mean-path (`chain * draw`) vs predictive (`chain * draw * unit`) sample lengths.

`make test-fast` (891) and `make lint` clean.

File: pathmc/simulate.py [hunk 1]
```diff
@@ -149,40 +329,27 @@ def time_index(self) -> np.ndarray | None:
...
-
-        return DoResult(
-            values=new_values,
-            values_by_time=new_by_time,
-            time_index=self._time_index,
-        )
+        # Align on shared variables; xarray handles dim alignment.
+        common = [v for v in self._ds.data_vars if v in other._ds.data_vars]
+        diff = self._ds[common] - other._ds[common]
+        return DoResult(ds=diff)
```
[the remark this answers is not in the document]
Comment (drbenvincent) [reply pathmc/simulate.py]: Fixed in 6e23ae3. `__sub__` on both `DoResult` and `EstimandResult` now goes through `_subtract_stores()`, which stacks each operand to a flat `(n_samples,)` vector and subtracts positionally in numpy — same contract as the pre-xarray code. Mismatched lengths raise `ValueError` instead of xarray silently inner-joining or NaN-filling.

Added `test_subtract_legacy_vs_labeled_layout` (same flat draws, `chain`/`draw` vs `sample_{var}` internal layout → zero contrast) and `test_subtract_raises_on_length_mismatch`.

**Comment by drbenvincent** [comment]
Added public `dataset` property on `DoResult` and `EstimandResult` (f854715).

- `result.dataset` returns the labelled `xr.Dataset` (dims `chain`, `draw`, optional `unit` / `time`)
- Backing store remains `_ds`; `dataset` is a documented read-only property (ArviZ-style naming)
- `draws()` / `mean()` / `hdi()` stay the stable flat-numpy API; `dataset` is the power-user escape hatch
- `from_contrast` now passes `contrast.dataset`; sharing semantics unchanged
- Tests updated to assert via `dataset`; added `test_dataset_aliases_internal_store`
- Repr footers mention `.dataset`

**Comment by drbenvincent** [comment]
Follow-up commit **b7cc9c5** — tier 1+2 simplifications on top of the xarray migration:

**`_build_dataset` tightened**
- Flat dict constructors now require explicit `n_chains`/`n_draws` (raises otherwise).
- Removed `sample_{var}` legacy fallback and auto-inference of chain/draw from array length.
- Removed heterogeneous `unit_{var}` dims — predictive arrays always use shared `unit` via `n_units_per_var`.
- Dropped scalar `n_units` kwarg from `DoResult.__init__`.
- Unrecognised array lengths raise `ValueError` instead of silently degrading.

**Removed migration shims**
- Deleted `_values` / `_values_by_time` dict-view properties from both result classes (~65 lines).
- `by_time()` checks the dataset for a `time` dim directly (`_has_by_time`).
- `DoResult` reprs iterate `dataset.data_vars` + `_draw()` instead of `_values`.
- `_model.py`: `sensitivity()` uses `_draw(outcome)`; `prob()` builds namespace from `dataset.data_vars`.

**Contrast subtraction**
- `__sub__` inlines xarray subtraction (`lhs[common] - rhs[common]`); removed `_subtract_stores` helper.

**Tests**
- Fixtures updated with explicit chain/draw metadata.
- Removed tests for deleted APIs (`_values` dict view, heterogeneous units, legacy fallback).
- Added `TestBuildDatasetRequiresChainDraw`.

Net **−65 lines** in this commit. Tier 3 (producers emit `Dataset` directly, narrow constructors) tracked in #353.

`make lint` + `make test-fast` green.
